### PR TITLE
Remove `nil` check for ctx

### PR DESCRIPTION
**Pull request by MustafaSaber**
fixes https://github.com/zalando/skipper/issues/2136
Ran tests local with check removed and worked fine.

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

### CODE REVIEW

File: ratelimit/redis_test.go [hunk 1]
```diff
@@ -377,7 +376,7 @@ func TestFailOpenOnRedisError(t *testing.T) {
 		settings.Group,
 	)
 
-	allow := c.AllowContext(context.Background(), "akey")
```
Comment (AlexanderYastrebov) [inline ratelimit/redis_test.go:380]: Lets revert this change, Allow() is kind of deprecated actually

Comment (MustafaSaber) [reply ratelimit/redis_test.go:380]: It's used in all tests in this file actually

Comment (AlexanderYastrebov) [reply ratelimit/redis_test.go:380]: Then lets update tests - ratelimit filters call AllowContext

File: ratelimit/redis.go [hunk 2]
```diff
@@ -174,7 +174,19 @@ func (c *clusterLimitRedis) allow(ctx context.Context, clearText string) (bool,
...
+// skippers should be allowed else false. It will share it's own data
+// and use the current cluster information to calculate global rates
+// to decide to allow or not.
+//
+// Performance considerations:
+//
+// In case of deny it will use ZREMRANGEBYSCORE and ZCARD commands in
+// one pipeline to remove old items in the list of hits.
+// In case of allow it will additionally use ZADD with a second
+// roundtrip.
```
Comment (AlexanderYastrebov) [inline ratelimit/redis.go:187]: Lets keep previous comment, no need to draw reader attention and later tell them its deprecated :)

File: ratelimit/swim.go [hunk 3]
```diff
@@ -78,10 +79,44 @@ func newClusterRateLimiterSwim(s Settings, sw Swarmer, group string) *clusterLim
...
+	log.Debugf("%s clusterRatelimit: Allow=%v, %v < %d", c.group, result, rate, c.maxHits)
+	return result
+}
+
 // Allow returns true if the request calculated across the cluster of
 // skippers should be allowed else false. It will share it's own data
 // and use the current cluster information to calculate global rates
 // to decide to allow or not.
+//
+// Deprecated: In favour of AllowContext
```
Comment (szuecs) [inline ratelimit/swim.go:119]: I think what you want to do in Allow() is to call `AllowContext(context.Background(), ..)`

File: ratelimit/ratelimit.go [hunk 4]
```diff
@@ -362,6 +368,8 @@ type limiter interface {
 
 // contextLimiter extends limiter with an AllowContext method that accepts an additional
 // context.Context, e.g. to support OpenTracing.
+//
+// Deprecated: In favour of limiter since Allow is also deprecated now.
```
Comment (szuecs) [inline ratelimit/ratelimit.go:372]: Can we clean this up?
We don't need contextLimiter if it does not differ from limiter interface and it's internal only so we can do whatever we want.

File: ratelimit/ratelimit.go [hunk 5]
```diff
@@ -338,8 +338,14 @@ func (s Settings) String() string {
 type limiter interface {
 	// Allow is used to get a decision if you should allow the
 	// call to pass or to ratelimit
+	//
+	// Depreceated
```
Comment (szuecs) [inline ratelimit/ratelimit.go:342]: Depreceated -> Deprecated

Comment (MustafaSaber) [reply ratelimit/ratelimit.go]: I didn't delete it but if you like I can, I left it in favour of deprecation first then deletion.

Comment (szuecs) [reply ratelimit/ratelimit.go]: yes please delete

Comment (MustafaSaber) [reply ratelimit/ratelimit.go]: From implementations also or just the interface?

File: go.mod [hunk 6]
```diff
@@ -33,8 +33,8 @@ require (
 	github.com/sarslanhan/cronmask v0.0.0-20190709075623-766eca24d011
 	github.com/sirupsen/logrus v1.8.1
 	github.com/sony/gobreaker v0.5.0
-	github.com/stretchr/testify v1.8.0
-	github.com/szuecs/rate-limit-buffer v0.7.1
+	github.com/stretchr/testify v1.7.0
```
Comment (AlexanderYastrebov) [inline go.mod:36]: Downgrade?

File: go.mod [hunk 7]
```diff
@@ -33,8 +33,8 @@ require (
...
+	github.com/szuecs/rate-limit-buffer v0.7.3-0.20221230130449-bee752deeb50
```
Comment (AlexanderYastrebov) [inline go.mod:37]: Should we perform lib release to get v0.7.3 maybe?

Comment (MustafaSaber) [reply go.mod:37]: I think yes we should even increase the minor version because of the breaking changes, @szuecs wdyt?

Comment (szuecs) [reply go.mod:37]: v0.8.0 here you go

File: ratelimit/ratelimit.go [hunk 8]
```diff
@@ -336,9 +336,9 @@ func (s Settings) String() string {
 
 // limiter defines the requirement to be used as a ratelimit implmentation.
 type limiter interface {
-	// Allow is used to get a decision if you should allow the
-	// call to pass or to ratelimit
-	Allow(string) bool
+	// AllowContext is used to get a decision if you should allow the
+	// call with context to pass or to ratelimit
+	AllowContext(context.Context, string) bool
```
Comment (AlexanderYastrebov) [inline ratelimit/ratelimit.go:341]: Do we need "Context" in the name? 
Maybe it could be just `Allow(context.Context, string) bool`, wdyt?

@szuecs ?

Comment (AlexanderYastrebov) [reply ratelimit/ratelimit.go:341]: also I think we do not need "with context" addition in the docs - method obviously requires context

File: ratelimit/swim.go [hunk 9]
```diff
@@ -78,11 +79,11 @@ func newClusterRateLimiterSwim(s Settings, sw Swarmer, group string) *clusterLim
 	return rl
 }
 
-// Allow returns true if the request calculated across the cluster of
+// AllowContext returns true if the request calculated across the cluster of
 // skippers should be allowed else false. It will share it's own data
 // and use the current cluster information to calculate global rates
-// to decide to allow or not.
-func (c *clusterLimitSwim) Allow(clearText string) bool {
+// to decide to allow or not. With additional context.Context
```
Comment (AlexanderYastrebov) [inline ratelimit/swim.go:85]: "With additional context.Context" is redundant - context is required by this method

File: ratelimit/ratelimit.go (hunk 8 above)
Comment (szuecs) [reply ratelimit/ratelimit.go:341]: Yes let's go with `Allow(context.Context, string) bool`

File: ratelimit/swim.go (hunk 3 above)
Comment (szuecs) [reply ratelimit/swim.go]: Let's drop AllowContext and just add contect to Allow

File: ratelimit/ratelimit.go [hunk 10]
```diff
@@ -362,47 +362,22 @@ type limiter interface {
...
-// ratelimited
-func (l *Ratelimit) Allow(s string) bool {
-	if l == nil {
-		return true
-	}
-	return l.impl.Allow(s)
-}
-
-// AllowContext is like Allow but accepts an optional context.Context, e.g. to
+// Allow is like Allow but accepts an optional context.Context, e.g. to
```
Comment (AlexanderYastrebov) [inline ratelimit/ratelimit.go:372]: `Allow is like Allow` - so true :)

File: ratelimit/ratelimit.go [hunk 11]
```diff
@@ -362,47 +362,23 @@ type limiter interface {
...
-func (l *Ratelimit) Allow(s string) bool {
-	if l == nil {
-		return true
-	}
-	return l.impl.Allow(s)
-}
-
-// AllowContext is like Allow but accepts an optional context.Context, e.g. to
+// Allow is used to get a decision if you should allow the call and
+// accepts an optional context.Context, e.g. to
```
Comment (AlexanderYastrebov) [inline ratelimit/ratelimit.go:373]: Please check all godoc comments - context is not optional anymore

**Comment by AlexanderYastrebov** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:
